pulp_2to3_migration/app/plugin/content.py

```diff
@@ -201,55 +201,74 @@ def get_remote_by_importer_id(importer_id):
-                    remote=None,
-                    deferred_download=False)
-                dc = DeclarativeContent(content=pulp3content, d_artifacts=[da])
-                dc.extra_data = future_relations
-                await self.put(dc)
-
-            # Downloaded or on_demand content with LCEs.
-            #
-            # To create multiple remote artifacts, create multiple instances of declarative
-            # content which will differ by url/remote in their declarative artifacts
-            for lce in pulp2lazycatalog:
-                remote = get_remote_by_importer_id(lce.pulp2_importer_id)
-                deferred_download = not pulp2content.downloaded
-                if not remote and deferred_download:
-                    _logger.warn(_('On_demand content cannot be migrated without a remote '
-                                   'pulp2 unit_id: {}'.format(pulp2content.pulp2_id)))
+                if not pulp2content.downloaded and not pulp2lazycatalog:
+                    _logger.warn(_('On_demand content cannot be migrated without an entry in the '
+                                   'lazy catalog, pulp2 unit_id: '
+                                   '{}'.format(pulp2content.pulp2_id)))
                     continue
 
-                da = DeclarativeArtifact(
-                    artifact=artifact,
-                    url=lce.pulp2_url,
-                    relative_path=pulp_2to3_detail_content.relative_path_for_content_artifact,
-                    remote=remote,
-                    deferred_download=deferred_download)
-                dc = DeclarativeContent(content=pulp3content, d_artifacts=[da])
+            # create pulp3 content and assign relations if present
+            pulp3content, extra_info = await pulp_2to3_detail_content.create_pulp3_content()
+            future_relations = {'pulp2content': pulp2content}
+            if extra_info:
+                future_relations.update(extra_info)
+
+            # not all content units have files, create DC without artifact
+            if pulp_2to3_detail_content.pulp2_type in migrator.artifactless_types:
+                # dc without artifact
+                dc = DeclarativeContent(content=pulp3content)
                 dc.extra_data = future_relations
                 await self.put(dc)
+            else:
+                # create artifact for content that has file
+                # TODO potentially improve this part for content that can have multiple files
+                artifact = await self.create_artifact(pulp2content.pulp2_storage_path,
+                                                      pulp_2to3_detail_content.expected_digests,
+                                                      pulp_2to3_detail_content.expected_size,
+                                                      downloaded=pulp2content.downloaded)
+                
+
+                if pulp_2to3_detail_content.pulp2_type in lazy_types:
```

Comment:
@goosemania please double check the logic here, i might have misinterpreted things.